File: slicing-android/app/src/main/java/com/crescentflare/slicinggame/components/utility/ImageSource.kt
package com.crescentflare.slicinggame.components.utility

import android.content.Context
import android.content.res.Resources
import android.graphics.Bitmap
import android.graphics.drawable.Drawable
import androidx.core.content.ContextCompat
import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.bitmap_recycle.BitmapPool
import com.bumptech.glide.load.resource.bitmap.BitmapTransformation
import com.bumptech.glide.request.target.CustomTarget
import com.bumptech.glide.request.transition.Transition
import com.bumptech.glide.util.Util
import com.crescentflare.slicinggame.infrastructure.coreextensions.urlDecode
import com.crescentflare.slicinggame.infrastructure.coreextensions.urlEncode
import com.crescentflare.slicinggame.infrastructure.inflator.Inflators
import java.security.MessageDigest


/**
 * Component utility: defines the source of an image
 */
class ImageSource: Comparable<ImageSource> {

    // --
    // Static: factory method
    // --

    companion object {

        fun fromValue(value: Any?): ImageSource? {
            if (value is Map<*, *>) {
                val result: Map<String, Any>? = Inflators.viewlet.mapUtil.asStringObjectMap(value)
                if (result != null) {
                    return ImageSource(result)
                }
            } else {
                val tempMap = mapOf(Pair("value", value))
                Inflators.viewlet.mapUtil.optionalString(tempMap, "value", null)?.let {
                    return ImageSource(it)
                }
            }
            return null
        }

    }


    // --
    // Members
    // --

    var type = Type.Unknown
    var parameters = mutableMapOf<String, Any>()
    var name = ""


    // --
    // Initialization
    // --

    constructor(string: String) {
        // Extract type from scheme
        var checkString = string
        val schemeMarker = checkString.indexOf("://")
        if (schemeMarker >= 0) {
            type = Type.fromString(checkString.substring(0, schemeMarker))
            checkString = checkString.substring(schemeMarker + 3)
        }

        // Extract parameters
        val parameterMarker = checkString.indexOf('?')
        if (parameterMarker >= 0) {
            // Get parameter string
            val parameterString = checkString.substring(parameterMarker + 1)
            checkString = checkString.substring(0, parameterMarker)

            // Split into separate parameters and fill dictionary
            val parameterItems = parameterString.split("&")
            for (parameterItem in parameterItems) {
                val parameterSet = parameterItem.split("=")
                if (parameterSet.size == 2) {
                    val key = parameterSet[0].urlDecode()
                    parameters[key] = parameterSet[1].urlDecode()
                }
            }
        }

        // Finally set name to the remaining string
        name = checkString
    }

    constructor(map: Map<String, Any>) {
        val mapUtil = Inflators.viewlet.mapUtil
        type = Type.fromString(mapUtil.optionalString(map, "type", null))
        mapUtil.optionalString(map, "name", null)?.let {
            name = it
        }
        parameters = map.filter { it.key != "type" && it.key != "name" }.toMutableMap()
    }

    constructor(context: Context, resourceId: Int) {
        type = Type.InternalImage
        name = context.resources.getResourceEntryName(resourceId)
    }


    // --
    // Extract values
    // --

    private val density: Float?
        get() {
            if (parameters.containsKey("density")) {
                return Inflators.viewlet.mapUtil.optionalFloat(parameters, "density", 1f)
            }
            return null
        }

    private val forceWidth: Int?
        get() {
            val width = Inflators.viewlet.mapUtil.optionalDimension(parameters, "forceWidth", 0)
            if (width > 0) {
                return width
            }
            return null
        }

    private val forceHeight: Int?
        get() {
            val height = Inflators.viewlet.mapUtil.optionalDimension(parameters, "forceHeight", 0)
            if (height > 0) {
                return height
            }
            return null
        }


    // --
    // Conversion
    // --

    val uri: String
        get() {
            var uri = "${type.value}://$name"
            getParameterString()?.let {
                uri += "?$it"
            }
            return uri
        }

    val map: Map<String, Any>
        get() {
            val map = mutableMapOf<String, Any>(Pair("type", type.value), Pair("name", name))
            map.putAll(parameters)
            return map
        }


    // --
    // Obtain drawable
    // --

    fun getDrawable(context: Context, callback: (drawable: Drawable?) -> Unit) {
        if (type == Type.InternalImage) {
            val resourceId = context.resources.getIdentifier(name, "drawable", context.packageName)
            if (resourceId > 0) {
                return callback(ContextCompat.getDrawable(context, resourceId))
            } else {
                callback(null)
            }
        } else if (type == Type.Online || type == Type.SecureOnline) {
            Glide.with(context)
                .`as`(Drawable::class.java)
                .load(uri)
                .transform(BitmapScaler(forceWidth, forceHeight, density))
                .into(object : CustomTarget<Drawable>() {
                    override fun onLoadCleared(placeholder: Drawable?) {
                        // No implementation
                    }

                    override fun onLoadFailed(errorDrawable: Drawable?) {
                        callback(null)
                    }

                    override fun onResourceReady(resource: Drawable, transition: Transition<in Drawable>?) {
                        callback(resource)
                    }
                })
        } else {
            callback(null)
        }
    }


    // --
    // Comparable implementation
    // --

    override fun compareTo(other: ImageSource): Int {
        return uri.compareTo(other.uri)
    }


    // --
    // Helper
    // --

    private fun getParameterString(ignoreParams: List<String> = emptyList()): String? {
        if (parameters.isNotEmpty()) {
            var parameterString = ""
            for (key in parameters.keys.sorted()) {
                var ignore = false
                for (ignoreParam in ignoreParams) {
                    if (key == ignoreParam) {
                        ignore = true
                        break
                    }
                }
                val stringValue = Inflators.viewlet.mapUtil.optionalString(parameters, key, null)
                if (!ignore && stringValue != null) {
                    if (parameterString.isNotEmpty()) {
                        parameterString += "&"
                    }
                    parameterString += key.urlEncode() + "=" + stringValue.urlEncode()
                }
            }
            if (parameterString.isNotEmpty()) {
                return parameterString
            }
        }
        return null
    }


    // --
    // Type enum
    // --

    enum class Type(val value: String) {

        Unknown("unknown"),
        InternalImage("app"),
        Online("http"),
        SecureOnline("https");

        companion object {

            fun fromString(string: String?): Type {
                for (enum in values()) {
                    if (enum.value == string) {
                        return enum
                    }
                }
                return Unknown
            }

        }

    }


    // --
    // Image transformer class
    // --

    private class BitmapScaler(val forceWidth: Int? = null, val forceHeight: Int? = null, val density: Float? = null) : BitmapTransformation() {

        private val objectName = "ImageSource.BitmapScaler"

        override fun transform(pool: BitmapPool, toTransform: Bitmap, outWidth: Int, outHeight: Int): Bitmap {
            var newWidth = toTransform.width
            var newHeight = toTransform.height
            if (forceWidth != null) {
                newWidth = forceWidth
                newHeight = forceHeight ?: toTransform.height * newWidth / toTransform.width
            } else if (forceHeight != null) {
                newHeight = forceHeight
                newWidth = toTransform.width * newHeight / toTransform.height
            } else if (density != null && density != 0f) {
                newWidth = (toTransform.width.toFloat() * Resources.getSystem().displayMetrics.density / density).toInt()
                newHeight = (toTransform.height.toFloat() * Resources.getSystem().displayMetrics.density / density).toInt()
            }
            if (newWidth > 0 && newHeight > 0 && (newWidth != toTransform.width || newHeight != toTransform.height)) {
                return Bitmap.createScaledBitmap(toTransform, newWidth, newHeight, true)
            }
            return toTransform
        }

        override fun updateDiskCacheKey(messageDigest: MessageDigest) {
            messageDigest.update(objectName.toByteArray())
            if (forceWidth != null) {
                messageDigest.update(forceWidth.toString().toByteArray())
            }
            if (forceHeight != null) {
                messageDigest.update(forceHeight.toString().toByteArray())
            }
            if (density != null) {
                messageDigest.update(density.toString().toByteArray())
            }
        }

        override fun equals(other: Any?): Boolean {
            if (other is BitmapScaler) {
                return other.forceWidth == forceWidth && other.forceHeight == forceHeight && other.density == density
            }
            return false
        }

        override fun hashCode(): Int {
            val hashCodes = listOf(
                Util.hashCode(forceWidth ?: 0),
                Util.hashCode(forceHeight ?: 0),
                Util.hashCode(density ?: 0f)
            )
            var result = objectName.hashCode()
            for (hashCode in hashCodes) {
                result = Util.hashCode(result, hashCode)
            }
            return result
        }

    }

}

File: slicing-ios/SlicingGame/Components/Utility/ImageSource.swift
//
//  ImageSource.swift
//  Component utility: defines the source of an image
//

import UIKit
import UniLayout
import AlamofireImage

enum ImageSourceType: String {
    
    case unknown = "unknown"
    case appInternal = "app"
    case online = "http"
    case secureOnline = "https"
    
}

class ImageSource: Equatable {
    
    // --
    // MARK: Members
    // --

    var type = ImageSourceType.unknown
    var parameters = [String: Any]()
    var name = ""
    

    // --
    // MARK: Factory method
    // --
    
    class func fromValue(value: Any?) -> ImageSource? {
        if let dictionaryValue = value as? [String: Any] {
            return ImageSource(dictionary: dictionaryValue)
        } else if let stringValue = Inflators.viewlet.convUtil.asString(value: value) {
            return ImageSource(string: stringValue)
        }
        return nil
    }

    
    // --
    // MARK: Initialization
    // --
    
    init(string: String) {
        // Extract type from scheme
        var checkString = string
        if let schemeMarker = checkString.range(of: "://") {
            type = ImageSourceType(rawValue: String(checkString[..<schemeMarker.lowerBound])) ?? .unknown
            checkString = String(checkString[schemeMarker.upperBound...])
        }
        
        // Extract parameters
        if let parameterMarker = checkString.range(of: "?") {
            // Get parameter string
            let parameterString = String(checkString[parameterMarker.upperBound...])
            checkString = String(checkString[..<parameterMarker.lowerBound])
            
            // Split into separate parameters and fill dictionary
            let parameterItems = parameterString.split(separator: "&").map(String.init)
            for parameterItem in parameterItems {
                let parameterSet = parameterItem.split(separator: "=").map(String.init)
                if parameterSet.count == 2 {
                    let key = parameterSet[0].urlDecode()
                    parameters[key] = parameterSet[1].urlDecode()
                }
            }
        }
        
        // Finally set name to the remaining string
        name = checkString
    }
    
    init(dictionary: [String: Any]) {
        let convUtil = Inflators.viewlet.convUtil
        type = ImageSourceType(rawValue: convUtil.asString(value: dictionary["type"]) ?? "unknown") ?? .unknown
        if let name = convUtil.asString(value: dictionary["name"]) {
            self.name = name
        }
        parameters = dictionary.filter { $0.key != "type" && $0.key != "name" }
    }

    
    // --
    // MARK: Extract values
    // --

    private var density: CGFloat? {
        return parameters["density"] != nil ? CGFloat(Inflators.viewlet.convUtil.asFloat(value: parameters["density"]) ?? 1.0) : nil
    }

    private var forceWidth: CGFloat? {
        return Inflators.viewlet.convUtil.asDimension(value: parameters["forceWidth"])
    }

    private var forceHeight: CGFloat? {
        return Inflators.viewlet.convUtil.asDimension(value: parameters["forceHeight"])
    }

    
    // --
    // MARK: Conversion
    // --

    var uri: String {
        get {
            var uri = "\(type.rawValue)://\(name)"
            if let parameterString = getParameterString() {
                uri += "?" + parameterString
            }
            return uri
        }
    }
    
    var dictionary: [String: Any] {
        get {
            var dictionary: [String: Any] = ["type": type.rawValue, "name": name]
            for (key, value) in parameters {
                dictionary[key] = value
            }
            return dictionary
        }
    }
    
    
    // --
    // MARK: Obtain image
    // --

    func getImage(completion: @escaping (_ image: UIImage?) -> Void) {
        if type == .appInternal {
            let bundle = Bundle(for: ImageSource.self)
            completion(UIImage(named: name, in: bundle, compatibleWith: nil))
        } else if type == .online || type == .secureOnline {
            if let imageUrl = URL(string: uri) {
                UIImageView.af_sharedImageDownloader.download(URLRequest(url: imageUrl), filter: ImageSourceImageScaler(forceWidth: forceWidth, forceHeight: forceHeight, density: density), completion: { response in
                    completion(response.result.value)
                })
            } else {
                completion(nil)
            }
        } else {
            completion(nil)
        }
    }
    

    // --
    // MARK: Equatable implementation
    // --

    static func == (lhs: ImageSource, rhs: ImageSource) -> Bool {
        return lhs.uri == rhs.uri
    }
    
    
    // --
    // MARK: Helper
    // --

    private func getParameterString(ignoreParams: [String] = []) -> String? {
        if parameters.count > 0 {
            var parameterString = ""
            for key in parameters.keys.sorted() {
                var ignore = false
                for ignoreParam in ignoreParams {
                    if key == ignoreParam {
                        ignore = true
                        break
                    }
                }
                if !ignore, let value = Inflators.viewlet.convUtil.asString(value: parameters[key]) {
                    if parameterString.count > 0 {
                        parameterString += "&"
                    }
                    parameterString += key.urlEncode() + "=" + value.urlEncode()
                }
            }
            if !parameterString.isEmpty {
                return parameterString
            }
        }
        return nil
    }

}

fileprivate struct ImageSourceImageScaler: ImageFilter {
    
    private let forceWidth: CGFloat?
    private let forceHeight: CGFloat?
    private let density: CGFloat?
    
    public init(forceWidth: CGFloat? = nil, forceHeight: CGFloat? = nil, density: CGFloat? = nil) {
        self.forceWidth = forceWidth
        self.forceHeight = forceHeight
        self.density = density
    }
    
    public var filter: (Image) -> Image {
        return { image in
            var newWidth = image.size.width
            var newHeight = image.size.height
            if let forceWidth = self.forceWidth {
                newWidth = forceWidth
                newHeight = self.forceHeight ?? image.size.height * newWidth / image.size.width
            } else if let forceHeight = self.forceHeight {
                newHeight = forceHeight
                newWidth = image.size.width * newHeight / image.size.height
            } else if let density = self.density, density != 0 {
                newWidth = image.size.width * UIScreen.main.scale / density
                newHeight = image.size.height * UIScreen.main.scale / density
            }
            if newWidth > 0 && newHeight > 0 && (newWidth != image.size.width || newHeight != image.size.height) {
                return image.af_imageScaled(to: CGSize(width: newWidth, height: newHeight))
            }
            return image
        }
    }
    
}
